### ISSUE TITLE
How can I use this program?

### ISSUE THREAD

**Issue by jergusg** [maxim-saplin/NetCoreStorageSpeedTest#1]
How to use this program on linux? Is there any manual or help? When I run the program It outputs:
```
CPDT_Console - STORAGE SPEED TEST

Total RAM: 0.00Gb, Available RAM: 0.00Gb

The test uses standrd OS's file API (WinAPI on Windows and POSIX on Mac) to
measure the speed of transfer between storage device and system memory.
 Write buffering is disabled

[ ] /boot/efi 0.06 Gb free - insufficient free space
- please pick drive to test: 
```
How can I pick drive? I tried entering path like /home/user/ but nothing works.

**Comment by maxim-saplin** [comment]
The app didn't find any partitions besides the boot partition (which it apparently can't write to). What OS do you use?

**Comment by maxim-saplin** [comment]
Also, could you please list partitions available in the system under same sandbox you might be using for the app? E.g. via` fdisk -l`

I'm not a linux expert, but can assume there can be some privileges issues.

**Comment by jergusg** [comment]
I use Ubuntu 19.10 and 18.04. I tried it on both, directly in system console.
This is output of sudo fdisk -l
```
dave@dave-ThinkPad:~/Downloads/CPDT_Console_Linux$ sudo fdisk -l
Disk /dev/loop0: 87,5 MiB, 91725824 bytes, 179152 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop1: 54,4 MiB, 57069568 bytes, 111464 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop2: 4,2 MiB, 4403200 bytes, 8600 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop3: 136,9 MiB, 143523840 bytes, 280320 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop4: 149,9 MiB, 157184000 bytes, 307000 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop5: 14,8 MiB, 15458304 bytes, 30192 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop6: 956 KiB, 978944 bytes, 1912 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop7: 3,7 MiB, 3821568 bytes, 7464 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/sda: 465,8 GiB, 500107862016 bytes, 976773168 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: CD8B9FC2-8696-4A93-AFFF-C771F8879572

Device         Start       End   Sectors   Size Type
/dev/sda1       2048    923647    921600   450M Windows recovery environment
/dev/sda2     923648   1128447    204800   100M EFI System
/dev/sda3    1128448   1161215     32768    16M Microsoft reserved
/dev/sda4    1161216 237869292 236708077 112,9G Microsoft basic data
/dev/sda5  237871104 239491071   1619968   791M Windows recovery environment
/dev/sda6  239493120 956291071 716797952 341,8G Microsoft basic data


Disk /dev/sdb: 119,2 GiB, 128035676160 bytes, 250069680 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: 6520C194-9824-4157-ADBE-48C82C487841

Device         Start       End   Sectors  Size Type
/dev/sdb1       2048 117233663 117231616 55,9G Linux filesystem
/dev/sdb2  117233664 250068991 132835328 63,3G Microsoft basic data


Disk /dev/sdc: 14,9 GiB, 16013942784 bytes, 31277232 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: gpt
Disk identifier: F2F63574-7940-4D14-94D3-A0C647DCCBA4

Device     Start      End  Sectors  Size Type
/dev/sdc1   2048 31277198 31275151 14,9G Linux filesystem


Disk /dev/mmcblk0: 236,1 GiB, 253545676800 bytes, 495206400 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x0022eb54

Device         Boot Start       End   Sectors   Size Id Type
/dev/mmcblk0p1 *     2048 495206399 495204352 236,1G  7 HPFS/NTFS/exFAT


Disk /dev/loop8: 89,1 MiB, 93429760 bytes, 182480 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop9: 44,2 MiB, 46325760 bytes, 90480 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop10: 42,8 MiB, 44879872 bytes, 87656 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop11: 88,5 MiB, 92778496 bytes, 181208 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop12: 84,5 MiB, 88629248 bytes, 173104 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop13: 3,7 MiB, 3825664 bytes, 7472 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop14: 156,7 MiB, 164290560 bytes, 320880 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop15: 4 MiB, 4218880 bytes, 8240 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop16: 1008 KiB, 1032192 bytes, 2016 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop17: 54,5 MiB, 57151488 bytes, 111624 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop18: 16 MiB, 16769024 bytes, 32752 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop19: 28,9 MiB, 30306304 bytes, 59192 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes


Disk /dev/loop20: 14,8 MiB, 15462400 bytes, 30200 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
```

**Comment by maxim-saplin** [comment]
Thanks. I can reproduce the problem, will look into it soon.

**Comment by maxim-saplin** [comment]
Could you please try this version?
https://github.com/maxim-saplin/NetCoreStorageSpeedTest/releases/download/2.0.1/CPDT_Console.zip

**Comment by jergusg** [comment]
Thanks, I tried it. It works but there are some problems.
If I choose any drive different from root I get error:
```
dave@dave-ThinkPad:~/Desktop/Downloads/CPDT_Console$ ./CPDT_Console 
CPDT_Console - STORAGE SPEED TEST

Total RAM: 0.00Gb, Available RAM: 0.00Gb

The test uses standrd OS's file API (WinAPI on Windows and POSIX on Mac/Linux)
to measure the speed of transfers between storage device and system memory.
 Write buffering and mem cahce are disabled

[ ] /sys 0.00 Gb free - insufficient free space
[ ] /proc 0.00 Gb free - insufficient free space
[1] /dev 3.80 Gb free 
[ ] /dev/pts 0.00 Gb free - insufficient free space
[ ] /run 0.76 Gb free - insufficient free space
[2] / 20.97 Gb free 
[3] /dev/shm 3.83 Gb free 
[ ] /run/lock 0.00 Gb free - insufficient free space
[ ] /dev/mqueue 0.00 Gb free - insufficient free space
[ ] /dev/hugepages 0.00 Gb free - insufficient free space
[4] /mnt/doghouse 153.36 Gb free 
[ ] /boot/efi 0.06 Gb free - insufficient free space
[ ] /run/snapd/ns 0.76 Gb free - insufficient free space
[ ] /run/snapd/ns/node-red.mnt 0.00 Gb free - insufficient free space
[ ] /run/user/1000 0.77 Gb free - insufficient free space
[ ] /run/user/1000/gvfs 0.00 Gb free - insufficient free space
[ ] /run/user/1000/doc 0.00 Gb free - insufficient free space
[5] /media/dave/Lexar 18.50 Gb free 
- please pick drive to test: 5

Program interupted due to unexpected error:
	System.ArgumentNullException Path cannot be null. (Parameter 'path')
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at Saplin.StorageSpeedMeter.TestFile..ctor(String drivePath, Int64 testAreaSizeBytes, Boolean writeBuffering, Boolean enableMemCache, String filePath, Action flush, Boolean mockStream, Boolean disableMacStream)
   at Saplin.StorageSpeedMeter.BigTest..ctor(String drivePath, Int64 fileSize, Boolean writeBuffering, MemCacheOptions memCache, String filePath, ICachePurger purger, String purgingFilePath, Func`1 freeMem, Int64 totalMem, WriteBufferFlusher flusher, Boolean mockFileStream, Boolean disableMacStream)
   at WinMacDiskSpeedTest.Program.Main(String[] args)

Press any key to quit
```
If I choose root, test starts:
```
[ ] /sys 0.00 Gb free - insufficient                           Elapsed: 00m 09s
[ ] /proc 0.00 Gb free - insufficie                            Elapsed: 00m 10s
[1] /dev 3.80 Gb free                                          Elapsed: 00m 16s
[ ] /dev/pts 0.00 Gb free - insuffi                          Remaining: 00m 14s
[ ] /run 0.76 Gb free - insufficient                           Elapsed: 00m 24s
[2] / 20.97 Gb free                                          Remaining: 00m 00s
[3] /dev/shm 3.83 Gb free                                      Elapsed: 00m 31s
[ ] /run/lock 0.00 Gb free - insuff                          Remaining: 00m 00s
[ ] /dev/mqueue 0.00 Gb free - insuff                          Elapsed: 00m 35s
[ ] /dev/hugepages 0.00 Gb free - i                          Remaining: 00m 00s
[4] /mnt/doghouse 153.36 Gb free 
[ ] /boot/efi 0.06 Gb free - insufficient free space
[ ] /run/snapd/ns 0.76 Gb free - insufficient free space
[ ] /run/snapd/ns/node-red.mnt 0.00 Gb free - insufficient free space
[ ] /run/user/1000 0.77 Gb free - insufficient free space
[ ] /run/user/1000/gvfs 0.00 Gb free - insufficient free space
[ ] /run/user/1000/doc 0.00 Gb free - insufficient free space
[5] /media/dave/Lexar 18.50 Gb free 
- please pick drive to test: 2
Test file: ~, Size: 1.00Gb

 Press ESC to break

1/5 Sequential write [4MB] block        Avg: 187.0MB/s   Min÷Max: 127.2 ÷ 219.6, Time: 0m09s
2/5 Sequential read [4MB] block         Avg: 2135.1MB/s  Min÷Max: 1873.2 ÷ 2847.1, Time: 0m06s
3/5 Random write [4KB] block            Avg: 1.9MB/s     Min÷Max: 0.8 ÷ 3.4, Time: 0m07s
4/5 Random read [4KB] block             Avg: 827.3MB/s   Min÷Max: 361.3 ÷ 1624.2, Time: 0m06s
5/5 Memory copy [4MB] block             Avg: 3726.7MB/s  Min÷Max: 3384.1 ÷ 4505.5, Time: 0m04s

Write Score*:	 8.55 MB/s
Read Score*:	 1549.53 MB/s
*Calculation: average throughput with 80% read/written seqentialy and 20% randomly

Test file deleted.  Saving results to CSV files in folder: StorageSpeedTestResults

Press any key to quit
```
However, the speeds are not real. If I try the same disk on windows, I get real values:
![Annotation 2020-02-02 120833](https://user-images.githubusercontent.com/7312449/73607396-dc8f5700-45b5-11ea-893d-c85c2cc985b4.png)

**Comment by maxim-saplin** [comment]
This one?
https://www.dropbox.com/s/bml7ena65iqg0pu/CPDT_Console.zip?dl=0

**Comment by jergusg** [comment]
Now it works with other drives. This is example of results:
```
dave@dave-ThinkPad:~/Downloads/Dropbox_CPDT/CPDT_Console$ sudo ./CPDT_Console                                                            Remaining: 00m 00s
CPDT_Console - STORAGE SPEED TEST

Total RAM: 0.00Gb, Available RAM: 0.00Gb

The test uses standrd OS's file API (WinAPI on Windows and POSIX on Mac/Linux) to measure the speed of transfers between storage device and system memory.
 Write buffering and mem cahce are disabled

[ ] /sys 0.00 Gb free - insufficient free space
[ ] /proc 0.00 Gb free - insufficient free space
[1] /dev 3.80 Gb free 
[ ] /dev/pts 0.00 Gb free - insufficient free space
[ ] /run 0.76 Gb free - insufficient free space
[2] / 20.95 Gb free 
[3] /dev/shm 3.72 Gb free 
[ ] /run/lock 0.00 Gb free - insufficient free space
[ ] /dev/mqueue 0.00 Gb free - insufficient free space
[ ] /dev/hugepages 0.00 Gb free - insufficient free space
[ ] /boot/efi 0.06 Gb free - insufficient free space
[ ] /run/snapd/ns 0.76 Gb free - insufficient free space
[ ] /run/snapd/ns/node-red.mnt 0.00 Gb free - insufficient free space
[ ] /run/user/1000 0.77 Gb free - insufficient free space
[ ] /run/user/1000/gvfs 0.00 Gb free - insufficient free space
[ ] /run/user/1000/doc 0.00 Gb free - insufficient free space
[4] /media/dave/Lexar 18.50 Gb free 
[5] /media/dave/d03764a2-fc25-4d47-b9bb-c6ba41cf2e27 5.49 Gb free 
- please pick drive to test: 5
Test file: /media/dave/d03764a2-fc25-4d47-b9bb-c6ba41cf2e27/CPDT_TestFile.dat, Size: 1.00Gb

 Press ESC to break

1/5 Sequential write [4MB] block        Avg: 38.2MB/s    Min÷Max: 19.6 ÷ 44.5, Time: 0m32s                                                                 
2/5 Sequential read [4MB] block         Avg: 2167.3MB/s  Min÷Max: 1751.5 ÷ 2810.8, Time: 0m06s                                                             
3/5 Random write [4KB] block            Avg: 1.5MB/s     Min÷Max: 0.3 ÷ 4.3, Time: 0m07s                                                                   
4/5 Random read [4KB] block             Avg: 631.3MB/s   Min÷Max: 271.5 ÷ 1208.2, Time: 0m06s                                                              
5/5 Memory copy [4MB] block             Avg: 3998.3MB/s  Min÷Max: 3504.5 ÷ 5293.6, Time: 0m04s                                                             

Write Score*:	 5.92 MB/s
Read Score*:	 1403.36 MB/s
*Calculation: average throughput with 80% read/written seqentialy and 20% randomly

Test file deleted.  Saving results to CSV files in folder: /media/dave/d03764a2-fc25-4d47-b9bb-c6ba41cf2e27/StorageSpeedTestResults

Press any key to quit
```
But for some drives it gives error:
```
dave@dave-ThinkPad:~/Downloads/Dropbox_CPDT/CPDT_Console$ sudo ./CPDT_Console                                                            Remaining: 00m 00s
CPDT_Console - STORAGE SPEED TEST

Total RAM: 0.00Gb, Available RAM: 0.00Gb

The test uses standrd OS's file API (WinAPI on Windows and POSIX on Mac/Linux) to measure the speed of transfers between storage device and system memory.
 Write buffering and mem cahce are disabled

[ ] /sys 0.00 Gb free - insufficient free space
[ ] /proc 0.00 Gb free - insufficient free space
[1] /dev 3.80 Gb free 
[ ] /dev/pts 0.00 Gb free - insufficient free space
[ ] /run 0.76 Gb free - insufficient free space
[2] / 20.95 Gb free 
[3] /dev/shm 3.72 Gb free 
[ ] /run/lock 0.00 Gb free - insufficient free space
[ ] /dev/mqueue 0.00 Gb free - insufficient free space
[ ] /dev/hugepages 0.00 Gb free - insufficient free space
[ ] /boot/efi 0.06 Gb free - insufficient free space
[ ] /run/snapd/ns 0.76 Gb free - insufficient free space
[ ] /run/snapd/ns/node-red.mnt 0.00 Gb free - insufficient free space
[ ] /run/user/1000 0.77 Gb free - insufficient free space
[ ] /run/user/1000/gvfs 0.00 Gb free - insufficient free space
[ ] /run/user/1000/doc 0.00 Gb free - insufficient free space
[4] /media/dave/Lexar 18.50 Gb free 
[5] /media/dave/d03764a2-fc25-4d47-b9bb-c6ba41cf2e27 5.48 Gb free 
- please pick drive to test: 4
Test file: /media/dave/Lexar/CPDT_TestFile.dat, Size: 1.00Gb

 Press ESC to break

1/5 Sequential write [4MB] block        Avg: 20.2MB/s    Min÷Max: 4.5 ÷ 37.4, Time: 0m56s                                                                  
2/5 Sequential read [4MB] block         Avg: 2181.2MB/s  Min÷Max: 1866.7 ÷ 2606.0, Time: 0m05s                                                             
3/5 Random write [4KB] block            Avg: 0.4MB/s     Min÷Max: 0.1 ÷ 0.9, Time: 0m07s                                                                   
4/5 Random read [4KB] block             Avg: 617.0MB/s   Min÷Max: 263.2 ÷ 1212.4, Time: 0m06s                                                              
5/5 Memory copy [4MB] block             Avg: 3903.6MB/s  Min÷Max: 3540.0 ÷ 4973.7, Time: 0m04s                                                             

Write Score*:	 1.80 MB/s
Read Score*:	 1397.19 MB/s
*Calculation: average throughput with 80% read/written seqentialy and 20% randomly

Test file deleted.  Saving results to CSV files in folder: /media/dave/Lexar/StorageSpeedTestResults

Program interupted due to unexpected error:
	System.IO.DirectoryNotFoundException Could not find a part of the path '/media/dave/Lexar/StorageSpeedTestResults/Aggrgate-Results-2020-02-03T12:18:56.csv'.
   at Interop.ThrowExceptionForIoErrno(ErrorInfo errorInfo, String path, Boolean isDirectory, Func`2 errorRewriter)
   at Microsoft.Win32.SafeHandles.SafeFileHandle.Open(String path, OpenFlags flags, Int32 mode)
   at System.IO.FileStream.OpenHandle(FileMode mode, FileShare share, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.StreamWriter.ValidateArgsAndOpenPath(String path, Boolean append, Encoding encoding, Int32 bufferSize)
   at System.IO.StreamWriter..ctor(String path, Boolean append)
   at System.IO.File.CreateText(String path)
   at Saplin.StorageSpeedMeter.TestSuite.SaveAggregateResults(String fileName, String separator, String decimalSeparator, DateTime now)
   at Saplin.StorageSpeedMeter.TestSuite.ExportToCsv(String folderPath, Boolean saveAllDataPoints, Nullable`1 dateTime, String separator, String decimalSeparator)
   at WinMacDiskSpeedTest.Program.Main(String[] args)

Press any key to quit
```
Speeds are not correct. Sequential write a read is quite off.

**Comment by maxim-saplin** [comment]
Spent some time cleaning the app and it seems that read caching on Linux is finally resolved, there's less clutter and no exceptions at the end of test.
https://github.com/maxim-saplin/NetCoreStorageSpeedTest/releases/download/2.0.2/CPDT_Console.zip

Regarding consistency of results of same device on different platforms, for my USB flash drive I did get same results on Mac and Ubuntu VM for sequential tests, though had significant difference for random tests which I don't have explanation yet.

P.S.: thanks for the effort troubleshooting the app.
